### ISSUE TITLE
AP Adapter - basic security audit and fixes

### ISSUE THREAD

**Issue by jjl** [bonfire-networks/bonfire-app#230]
I shall prefix this by saying I actually know *very little* about the AP adapter. I've fixed some code in it in the last few days, but I don't have a good high level understanding of how it all fits together, or indeed about the flows of ActivityPub in general beyond a basic overview.

I have noticed that a lot of AP queries do not appear to properly use boundaries. In my head there should probably be a clear distinction between unauthenticated and authenticated requests when passing things through to the adapter, but I do not know enough about how it all fits together to be sure of that.

Essentially, we would like to end up with:
* Unauthenticated queries being done as a guest.
* Authenticated queries being done as a user.
* Boundaries applied appropriately.

In short, ActivityPub should not be a hole through which you can do more than a local user.

**Comment by jjl** [comment]
@ajmeese7 do you think between us we could figure this one out?

**Comment by ajmeese7** [comment]
@jjl hey, sorry I've completely dropped off the radar. I've meant to help out more but I have been super busy these past few months, between surgery, college, taking on an NSA project lead position, and starting a company. I promise I still believe in you guys and I think this project is great!

I can definitely take a look, but I can't make any promises on whether or not I will be able to help. Unfortunately I also do not have much experience with AP, but I work in cybersecurity and have done a lot of API work in the past so I might be able to make something happen. Can you provide some examples of where in the code you mean so I can take a look?

**Comment by jjl** [comment]
Heh, sounds like you've been pretty busy! No worries.

The adapter behaviour from the `activity_pub` library is [here](https://github.com/bonfire-networks/activity_pub/blob/develop/lib/activity_pub/adapter.ex), everything tagged with `@callback`.

Our implementation of it is `bonfire_federate_activitypub`, [here](https://github.com/bonfire-networks/bonfire_federate_activitypub/blob/main/lib/adapter/adapter.ex).

I do not really understand the activitypub flows in detail - either just the specification or our implementation. My gut tells me that the adapter interface is not currently expressive enough to deal with identification. It was mostly built before we had a good idea of what bonfire's security systems would look like, so I guess that makes sense and this the painful integration part.

For the most part, I was hoping to make up for my lack of knowledge about AP with a second opinion. Oh well, I'll ask around :)

**Comment by ajmeese7** [comment]
You're right, that is a complicated problem to tackle. Unfortunately I am not yet very experienced with Elixir because I've been too bogged down with other projects to learn, so I won't be able to help on this quite yet.

Once everything else calms down (if it ever does) and I can get some practice with the language, I'll come back and try to help out. Sorry again, I hope you get it all figured out! I'll be watching the progress from a distance.

**Comment by jjl** [comment]
No worries, life happens :)

I spent some time yesterday reading a few sources about activitypub, including the specification and I think I found enough pertaining to the parts I was worried about from the audit perspective, and apparently I asked a specific enough question to ariadne that i got just "yes" in response, so that's progress!

In short, activitypub is a bit more limited than I had assumed, so for reading purposes, we should just assume everyone is a guest because only publishes have any sort of authentication.

**Comment by bhaugen** [comment]
@jjl @ajmeese7 @ivanminutillo @mayel 

I think you will find what you need somewhere in the https://spritelyproject.org/ from @cwebber and now also https://spritelyproject.org/news/interview-with-jessica-tallon.html

I think the main security-permission-authorization component is the "Capability Transport Protocol": https://spritelyproject.org/news/what-is-captp.html 

Christine was one of the founders of ActivityPub and has been working on Spritely ever since. She is also very helpful, as I also expect Jessica to be. They want this work to be used.

How much more work and headspace does Captp take beyond the usual ACLs?

**Comment by jjl** [comment]
> @jjl @ajmeese7 @ivanminutillo @mayel
> 
> I think you will find what you need somewhere in the https://spritelyproject.org/ from @cwebber and now also https://spritelyproject.org/news/interview-with-jessica-tallon.html
> 
> I think the main security-permission-authorization component is the "Capability Transport Protocol": https://spritelyproject.org/news/what-is-captp.html
> 
> Christine was one of the founders of ActivityPub and has been working on Spritely ever since. She is also very helpful, as I also expect Jessica to be. They want this work to be used.
> 

I reached out to christine a couple of days ago on twitter, but I didn't get a response yet. 

> How much more work and headspace does Captp take beyond the usual ACLs?

I have concerns about capabilities and compatibility with the rest of the fediverse that are definitely going to take longer to figure out than i have available for this work.

Mostly, i just needed enough context to figure out how authentication can be handled and tied in with the boundaries system, but i was surprised to discover that activitypub only seems to do any authentication in the real world on posting activities, so effectively all the reading should be done as guest anyway.

**Comment by jjl** [comment]
I'm declaring it to be good enough for alpha. I changed very little code.
